sycl/source/detail/graph_impl.cpp

```diff
@@ -1760,6 +1857,203 @@ void dynamic_parameter_base::updateAccessor(
+    // NewValue rather than assign it directly
+    std::memcpy(Arg.MPtr, NewValue, Size);
+    break;
+  }
+}
+
+void dynamic_parameter_impl::updateCGAccessor(
+    std::shared_ptr<sycl::detail::CG> &CG, int ArgIndex,
+    const sycl::detail::AccessorBaseHost *Acc) {
+  auto &Args = static_cast<sycl::detail::CGExecKernel *>(CG.get())->MArgs;
+
+  auto NewAccImpl = sycl::detail::getSyclObjImpl(*Acc);
+  for (auto &Arg : Args) {
+    if (Arg.MIndex != ArgIndex) {
+      continue;
+    }
+    assert(Arg.MType == sycl::detail::kernel_param_kind_t::kind_accessor);
+
+    // Find old accessor in accessor storage and replace with new one
+    if (static_cast<sycl::detail::SYCLMemObjT *>(NewAccImpl->MSYCLMemObj)
+            ->needsWriteBack()) {
+      throw sycl::exception(
+          make_error_code(errc::invalid),
+          "Accessors to buffers which have write_back enabled "
+          "are not allowed to be used in command graphs.");
+    }
+
+    // All accessors passed to this function will be placeholders, so we must
+    // perform steps similar to what happens when handler::require() is
+    // called here.
+    sycl::detail::Requirement *NewReq = NewAccImpl.get();
+    if (NewReq->MAccessMode != sycl::access_mode::read) {
+      auto SYCLMemObj =
+          static_cast<sycl::detail::SYCLMemObjT *>(NewReq->MSYCLMemObj);
+      SYCLMemObj->handleWriteAccessorCreation();
+    }
+
+    for (auto &Acc : CG->getAccStorage()) {
+      if (auto OldAcc = static_cast<sycl::detail::AccessorImplHost *>(Arg.MPtr);
+          Acc.get() == OldAcc) {
+        Acc = NewAccImpl;
+      }
+    }
+
+    for (auto &Req : CG->getRequirements()) {
+      if (auto OldReq = static_cast<sycl::detail::AccessorImplHost *>(Arg.MPtr);
+          Req == OldReq) {
+        Req = NewReq;
+      }
+    }
+    Arg.MPtr = NewAccImpl.get();
+    break;
+  }
+}
+
+dynamic_command_group_impl::dynamic_command_group_impl(
+    const command_graph<graph_state::modifiable> &Graph)
+    : MGraph{sycl::detail::getSyclObjImpl(Graph)}, MActiveCGF(0),
+      MNDRangeUsed(false) {}
+
+void dynamic_command_group_impl::finalizeCGFList(
+    const std::vector<std::function<void(handler &)>> &CGFList) {
+  // True if kernels use sycl::nd_range, and false if using sycl::range
+  bool NDRangeUsed = false;
+  for (size_t CGFIndex = 0; CGFIndex < CGFList.size(); CGFIndex++) {
+    const auto &CGF = CGFList[CGFIndex];
+    // Handler defined inside the loop so it doesn't appear to the runtime
+    // as a single command-group with multiple commands inside.
+    sycl::handler Handler{MGraph};
+    CGF(Handler);
+
+    if (Handler.getType() != sycl::detail::CGType::Kernel) {
+      throw sycl::exception(
+          make_error_code(errc::invalid),
+          "The only type of command-groups that can be used in "
+          "dynamic command-groups is kernels.");
+    }
+
+    Handler.finalize();
+
+    // Verify range/nd_range is consistent across CGs in list
```

Comment:
I think mixing nd_range and range should work in the backends. Is this restriction because of the interaction with update_nd_range?

If there is no backend limitations, should we have a task to relax this restriction? Also, don't think there is wording for this in the spec, so we might need to update it

Reply:
This restriction is because of the wording we currently have around `node::update_nd_range()` or `node::update_range()`. I updated the spec to include it.

update_range():
> Throws with error code invalid if the kernel node was originally created with a sycl::nd_range.

update_nd_range():
> Throws with error code invalid if the kernel node was originally created with a sycl::range.

If we the sycl::range(), sycl::nd_range() isn't consistent across all the command-groups in the node, then I don't think these errors make sense at the node level.  I'll have a look to see if we still need these restrictions because I think we've touched the UR spec about local size update since we first introduced this.

Reply:
Revisited this and yes we don't need to restrictions on mixing range/nd-range. We also don't need these restrictions any more on the `node::update_nd_range()` or `node::update_range()` update range APIs so i've removed these too. Here and in the spec PR.

Interestingly, i've kept a different in limitation between updating the dimensions of the range/nd_range. In the node level APIs you can't change the dimensions but in the dynamic command_group update you can change the dimensions. I think this makes sense because without updating the kernel code to a kernel defined in the different dimension, updating the dimension by itself doesn't make much sense. I've added spec wording to highlight this difference too.

Reply:
The changes look good to me. Yes, I agree that keeping the dimensions limitations at the node level makes sense.